lib/thread.c

```diff
@@ -1715,17 +1722,35 @@ void thread_call(struct thread *thread)
 				 memory_order_seq_cst);
 
 #ifdef CONSUMED_TIME_CHECK
-	if (realtime > CONSUMED_TIME_CHECK) {
+	if (cputime > CONSUMED_TIME_CHECK) {
 		/*
-		 * We have a CPU Hog on our hands.
+		 * We have a CPU Hog on our hands.  The time FRR
+		 * has spent doing actual work ( not sleeping )
+		 * is greater than 5 seconds.
 		 * Whinge about it now, so we're aware this is yet another task
 		 * to fix.
 		 */
+		atomic_fetch_add_explicit(&thread->hist->total_cpu_warn,
+					  1, memory_order_seq_cst);
 		flog_warn(
-			EC_LIB_SLOW_THREAD,
+			EC_LIB_SLOW_THREAD_CPU,
 			"SLOW THREAD: task %s (%lx) ran for %lums (cpu time %lums)",
 			thread->xref->funcname, (unsigned long)thread->func,
 			realtime / 1000, cputime / 1000);
+	} else if (realtime > CONSUMED_TIME_CHECK) {
+		/*
+		 * The runtime for a task is greater than 5 seconds, but
+		 * the cpu time is under 5 seconds.  Let's whine
+		 * about this because this could imply some sort of
+		 * scheduling issue.
+		 */
+		atomic_fetch_add_explicit(&thread->hist->total_wall_warn,
+					  1, memory_order_seq_cst);
+		flog_warn(
+			EC_LIB_SLOW_THREAD_WALL,
+			"SLOW THREAD: task %s (%lx) ran for %lums, with a cpu time of %lums",
```

Comment:
nit: lets use the same error message formatting used above, i.e. `(cpu time <blah>)` vs `with a cpu time of <blah>`. And let's add something to the start of each message (other than the error code value) that indicates whether we think this is starvation -vs- hog.

Reply:
fixed